src/index/crates.rs

```diff
@@ -0,0 +1,55 @@
+use crates_index::Crate;
+use failure::ResultExt;
+use std::io::{Seek, SeekFrom, Write};
+
+pub(crate) struct Crates {
+    repo: git2::Repository,
+}
+
+impl Crates {
+    pub(super) fn new(repo: git2::Repository) -> Self {
+        Self { repo }
+    }
+
+    pub(crate) fn walk(&self, mut f: impl FnMut(Crate)) -> Result<(), failure::Error> {
+        log::debug!("Walking crates in index");
+        let tree = self
+            .repo
+            .find_commit(self.repo.refname_to_id("refs/remotes/origin/master")?)?
+            .tree()?;
+
+        // crates_index doesn't publicly expose their slice constructor, so need to write each blob
+        // to a file before loading it as a `Crate`.
+        let mut tmp = tempfile::NamedTempFile::new()?;
+
+        let mut result = Ok(());
+
+        tree.walk(git2::TreeWalkMode::PreOrder, |_, entry| {
+            result = (|| {
+                if let Some(blob) = entry.to_object(&self.repo)?.as_blob() {
```

Comment:
This is getting the content, not the name.

Reply:
I don't know the git2 API well enough - can't you pass the name of the file to crates_index so you don't need a temporary file? Or does crates_index assume that the file is writable or something?

Reply:
The file doesn't actually exist anywhere, it is stored as part of git's compressed archives. That's why we read out the content and write it into a temporary named file that we can pass to `crates_index` to read back in and parse. I planned to take a look at whether it's possible to expose the [`crates_index::Crate::from_slice`](https://docs.rs/crates-index/0.15.0/src/crates_index/lib.rs.html#376-396) constructor so that we could just call that with the blob content directly (assuming that that's easier than changing `crates_index` to support bare git repos itself directly).

Reply:
I opened https://github.com/frewsxcv/rust-crates-index/issues/38, it _looks_ like they just have to add `pub` in front of the method.

Reply:
They just published a new version with the change :) https://crates.io/crates/crates-index/0.15.1